app/code/community/Bolt/Boltpay/OrderCreationException.php

```diff
@@ -0,0 +1,214 @@
+<?php
+/**
+ * Bolt magento plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2018 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+/**
+ * Class Bolt_Boltpay_OrderCreationException
+ *
+ * This exception is thrown when an error occurs while trying to create a backend order
+ */
+class Bolt_Boltpay_OrderCreationException extends Bolt_Boltpay_BoltException
+{
+
+    /**
+     * Database error or error in incoming request
+     */
+    const E_BOLT_GENERAL_ERROR               = 2001001; // verified
+    const E_BOLT_GENERAL_ERROR_TMPL_HMAC     = '{"reason": "Invalid HMAC header"}'; // verified
+    const E_BOLT_GENERAL_ERROR_TMPL_GENERIC  = '{"reason": "%s"}'; // verified
+
+    /**
+     * The order already exist in the system
+     */
+    const E_BOLT_ORDER_ALREADY_EXISTS        = 2001002; // verified
+    const E_BOLT_ORDER_ALREADY_EXISTS_TMPL   = '{"display_id": "%s", "order_status": "%s"}'; // verified
+
+    /**
+     * General non-item cart changes
+     */
+    const E_BOLT_CART_HAS_EXPIRED                         = 2001003;
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_EMPTY              = '{"reason": "Cart is empty"}';  // verified
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_EXPIRED            = '{"reason": "Cart has expired"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_NOT_FOUND          = '{"reason": "Cart does not exist with reference", "reference": "%s"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_NOT_PURCHASABLE    = '{"reason": "The product is not purchasable", "product_id": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_GRAND_TOTAL        = '{"reason": "Grand total has changed", "old_value": "%d", "new_value": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_DISCOUNT           = '{"reason": "Discount total has changed", "old_value": "%d", "new_value": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_TAX                = '{"reason": "Tax amount has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * Currently not used -- shipping tax in M1 is used to fix rounding problems and reports the
+     * full cart tax.  Therefore, checking full tax is sufficient
+     */
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_SHIPPING_TAX       = '{"reason": "Shipping tax amount has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * Item price changes
+     */
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED         = 2001004;
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED_TMPL    = '{"product_id": "%d", "old_price": "%d", "new_price": "%d"}';
```

Comment:
writing JSON with template variable seems strange to me. One downside is if you write an invalid JSON here you won't notice until runtime. How about we use plain PHP object or array and convert it to json string when this class generating error message?

Reply:
They are not a template _variables_.  They are appropriately unwavering template **constants**.  Plain old PHP objects can not be specified as constants.  Like with most programming languages, we are required to stick with scalar values for our use case.

The list JSON strings are all verified valid.  The function "__" which injects the variable portion of the templates uses `vsprintf` which assures the variables will be in the expected data type.  The design here is simple and solid.

The statement above is like saying, "if someone creates a PHP object that is encoded to JSON in an invalid format, we will not know until run time".  The solution is that should verify that the PHP object encodes to the expected format prior to PR.  Similarly, if someone adds to the constants or modifies them in the future, they should assure that they have written valid JSON before requesting a pull.